design/multi-client-and-multi-api-in-modular.md

```diff
@@ -0,0 +1,306 @@
+
+Let's say customers are using `@azure/foo/rest`, their user experience would be:
+
+```typescript
+// to use the helper function paginate
+ClientA.paginate(client, response);
+
+// to use the rest client type
+ClientA.ClientARestClient;
+
+// to access some models in client A
+ClientA.SomeModelInClientA;
+```
+
+And if customers are using `@azure/foo/rest/clientA`, their user experience would be:
+
+```typescript
+// to use the helper function paginate
+paginate(client, response);
+
+// to use the rest client type
+ClientARestClient;
+
+// to access some models in client A
+SomeModelInClientA;
+```
+
+In order to avoid this kind of user experience inconsistency and don't have to resolve the conflicts, as the `@azure/foo/rest/ClientA` and `@azure/foo/rest/clientB` belong to two different endpoints. we choose to remove the `src/rest/index.ts` and the subpath export `@azure/foo/rest`, just keep the sub clients `@azure/foo/rest/clientA` and `@azure/foo/rest/clientB`.
+
+## Finalized Proposals
+
+With the above initial proposals and the above considerations, we get our finalized design for both multi-client and multi-endpoint.
+
+### Finalized Proposal in Multi-Client
+
+<!-- markdownlint-disable MD033 -->
+<table>
+  <tr>
+    <th>Subpath Exports</th>
+    <th>Source Code Structure</th>
+  </tr>
+  <tr>
+    <td>
+      <pre lang="typescript">
+Both Exported
+@azure/foo
+@azure/foo/api
+@azure/foo/rest
+</pre>
+<pre lang="typescript">
+ClientA
+@azure/foo/clientA
+@azure/foo/clientA/api
+</pre>
+<pre lang="typescript">
+ClientB
+@azure/foo/clientB
+@azure/foo/clientB/api
+</pre>
+</td>
+<td>
+<pre lang="typescript">
+src
+src/api
+src/rest
+src/models
+// src/customize
+</pre>
+<pre lang="typescript">
+src/clientA
+src/clientA/api
+src/clientA/models
+// src/clientA/customize
+</pre>
+<pre lang="typescript">
+src/clientB
+src/clientB/api
+src/clientB/models
+// src/clientB/customize
+</pre>
+</td>
+  </tr>
+</table>
+<!-- markdownlint-enable MD033 -->
+
+### Finalized Proposal in Multi-Endpoint
+
+<!-- markdownlint-disable MD033 -->
+<table>
+  <tr>
+    <th>Subpath Exports</th>
+    <th>Source Code Structure</th>
+  </tr>
+  <tr>
+    <td>
+      <pre lang="typescript">
+Both Exported
+@azure/foo
+@azure/foo/api
+</pre>
+<pre lang="typescript">
+ClientA
+@azure/foo/clientA
+@azure/foo/clientA/api
+@azure/foo/rest/clientA
+</pre>
+<pre lang="typescript">
+ClientB
+@azure/foo/clientB
+@azure/foo/clientB/api
+@azure/foo/rest/clientB
+</pre>
+</td>
+<td>
+<pre lang="typescript">
+src
+src/api
+src/rest // without index.ts just sub folders in it
+src/models
+// src/customize
+</pre>
+<pre lang="typescript">
+src/clientA
+src/clientA/api
+src/clientA/models
+// src/clientA/customize
+src/rest/clientA
+</pre>
+<pre lang="typescript">
+src/clientB
+src/clientB/api
+src/clientB/models
+// src/clientB/customize
```

Comment:
Not sure the customization section is finalized, because I notice that it may not align with Jose's customization tools.